packages/app/src/cli/models/extensions/specifications/type-generation.ts

```diff
@@ -0,0 +1,218 @@
+          resolvedPaths.push(resolvedPath)
+        }
+      } else {
+        // Fallback to manual resolution for edge cases
+        // eslint-disable-next-line no-await-in-loop
+        const fallbackPath = await fallbackResolve(importPath, dirname(filePath))
+        if (fallbackPath) {
+          resolvedPaths.push(fallbackPath)
+        }
+      }
+    }
+
+    return resolvedPaths
+  } catch (error) {
+    // Re-throw AbortError as-is, wrap other errors
+    if (error instanceof AbortError) {
+      throw error
+    }
+    return []
+  }
+}
+
+export async function findAllImportedFiles(filePath: string, visited = new Set<string>()): Promise<string[]> {
+  if (visited.has(filePath)) {
+    return []
+  }
+
+  visited.add(filePath)
+  const resolvedPaths = await parseAndResolveImports(filePath)
+
+  const allFiles = [...resolvedPaths]
+
+  // Recursively find imports from the resolved files
+  for (const resolvedPath of resolvedPaths) {
+    // eslint-disable-next-line no-await-in-loop
+    const nestedImports = await findAllImportedFiles(resolvedPath, visited)
+    allFiles.push(...nestedImports)
+  }
+
+  return [...new Set(allFiles)]
+}
+
+export function createTypeDefinition(
+  fullPath: string,
+  typeFilePath: string,
+  targets: string[],
+  apiVersion: string,
+): string | null {
+  try {
+    // Validate that all targets can be resolved
+    for (const target of targets) {
+      try {
+        require.resolve(`@shopify/ui-extensions/${target}`, {paths: [fullPath, typeFilePath]})
+      } catch (_) {
+        const {year, month} = parseApiVersion(apiVersion) ?? {year: 2025, month: 10}
+        // Throw specific error for the target that failed, matching the original getSharedTypeDefinition behavior
+        throw new AbortError(
+          `Type reference for ${target} could not be found. You might be using the wrong @shopify/ui-extensions version.`,
+          `Fix the error by ensuring you have the correct version of @shopify/ui-extensions, for example ~${year}.${month}.0, in your dependencies.`,
+        )
+      }
+    }
+
+    const relativePath = relativizePath(fullPath, dirname(typeFilePath))
+
+    if (targets.length === 1) {
+      const target = targets[0] ?? ''
+      return `//@ts-ignore\ndeclare module './${relativePath}' {\n  const shopify: import('@shopify/ui-extensions/${target}').Api;\n  const globalThis: { shopify: typeof shopify };\n}\n`
+    } else if (targets.length > 1) {
+      const unionType = targets.map((target) => `    import('@shopify/ui-extensions/${target}').Api`).join(' |\n')
+      return `//@ts-ignore\ndeclare module './${relativePath}' {\n  const shopify: \n${unionType};\n  const globalThis: { shopify: typeof shopify };\n}\n`
+    }
+
+    return null
+  } catch (error) {
+    // Re-throw AbortError as-is, wrap other errors
+    if (error instanceof AbortError) {
+      throw error
+    }
+    const {year, month} = parseApiVersion(apiVersion) ?? {year: 2025, month: 10}
```

Comment:
Can you put this error handling into a share util instead of duplicating the code?

Reply:
The lint rules discourage returning an instance of abort-error hence the duplication 😅